### DISCUSSION TITLE
Why the target is x86_64-unknown-linux-musl?

### DISCUSSION THREAD

**Discussion by ghost** [aws/aws-lambda-rust-runtime#306] [Q&A] [answered]
Looks like the target is x86_64 archtecture. Could you explain on why the target is x86_64 archtecture?

Thanks!

**Comment by davidbarsky** [comment, marked as answer]
(Moved this to discussions instead of an issue).

Assuming that you're asking why we're using `x86_64-unknown-linux-musl` instead of `x86_64-unknown-linux-gnu`, it's because the `provided` environment had a glibc that's too old for Rust. Statically linking the musl libc is a good way to get around that.

With the launch of the `provided.al2` environment, the glibc _is_ recent enough for Rust, which means that if folks are targeting `provided.al2`, then they can use x86_64-unknown-linux-gnu as a target instead.

**Reply by ghost** [reply]
Got it.

I can use x86_64-unknown-linux-gnu as a target instead, if set the runtime to provide.al2.
Right?

**Reply by davidbarsky** [reply]
> I can use x86_64-unknown-linux-gnu as a target instead, if set the runtime to provide.al2. Right?

That's correct.

**Reply by seanpianka** [reply]
Is there any major differences in performance, behavior, or something else between musl and gnu?

**Reply by davidbarsky** [reply]
Before taking my advice, please run your own benchmarks. It's possible that my advice is out-of-date!

---

For some time, (and this still might be the case) musl's default allocator didn't perform well with idiomatic Rust code that made many small allocations across several threads. I _strongly_ suspect that's why customers found the cold-start times of this runtime to be slow. To mitigate this issue, I suggest that that customers should use an allocator like [jemalloc](https://github.com/gnzlbg/jemallocator) or [mimalloc](https://github.com/purpleprotocol/mimalloc_rust) when writing _any_ server-side (Lambda or otherwise) Rust code that uses the `x86_64-unknown-linux-musl` target.

With the `x86_64-unknown-linux-gnu` target, I believe that the allocation issue that afflicts musl-using Rust code is significantly less pronounced (see mimalloc's benchmark suite https://github.com/microsoft/mimalloc#benchmark-results-on-a-16-core-amd-5950x-zen3). However, I still think it's worthwhile trying out alternative allocators if you have any qualms.

In terms of the build/packing experience, I think that the two are largely the same.

**Reply by goshoo** [reply]
I found this blog post.

https://andygrove.io/2020/05/why-musl-extremely-slow/

**Reply by nk9** [reply]
Lots of info in [this blog post](https://nickb.dev/blog/default-musl-allocator-considered-harmful-to-performance/) about the default musl allocator w.r.t. Rust.
